### ISSUE TITLE
Proposal: Generic list/sequence interface

### ISSUE THREAD

**Issue by pdubroy** [cdglabs/ohm#63]
One of the nice things about the way we handle `_iter` nodes is that they do `map` for free. That is, `iterNode.interpret()` produces an Array which is the result of invoking the `interpret` operation on each of the children of the `_iter` node. With arrays, you'd have to write `anArray.map(function(x) { return x.interpret(); })`.

What if we take this idea, and make it more generally available in Ohm as a "List" or "Sequence" interface? 

In #62, I proposed adding a built-in `asArray` operation for ListOf nodes. It would return an Array of wrappers -- which is very similar to what `_iter` nodes do. Instead of `asArray`, maybe we should implement `asSequence`, so that you would have the same convenience of mapping that you get with `_iter` nodes.

Other places we could use this: the `children` property of nodes, and as the argument to `_default` and `_nonterminal` actions.

**Comment by mroeder** [comment]
I like this idea! I have always been wondering why the behavior of `ListOf` isn't like the `*` operator or `NonemptyListOf` like `+`. Almost always you don't care about the separator (as it of is a terminal like `,` etc. and nothing that varies) and so it could _just work_.

I don't understand the `children` example right but I generally think that everything that is _iterable_ should be handle like that per default. And you can always override it, of course.

**Comment by pdubroy** [comment]
I meant `this.node.children`, which is available on any wrapper object...but on second thought, scratch that because it is an array of CST nodes, not an array of wrappers.

It should still be applicable to the argument that is passed to `_nonterminal` though.

**Comment by pdubroy** [comment]
@alexwarth, any thoughts on this?

**Comment by pdubroy** [comment]
Ok, I've implemented the attribute and I think I am happy with it so far. You can see an example of using it at the top of this commit: https://github.com/cdglabs/ohm/commit/b320710044b2cc6a1509b73f5f8e40933fe54b43 

And here's what it looks like to implement `asSequence` for a user-defined rule: https://github.com/cdglabs/ohm/blob/master/test/test-semantics.js#L560. Because a Sequence implements the wrapper interface, you need a semantics instance to construct it (so that the sequence will have properties for all the operations/attributes in the semantics). For that reason, they are constructed by calling the `sequence` method on any wrapper instance.

**Comment by pdubroy** [comment]
I implemented passing a Sequence (instead of an Array) as the arg to `_default` and `_nonterminal`.

It basically means that you don't need to call `map`, `forEach`, etc. inside a `_default` or `_nonterminal` action. Some examples:

- https://github.com/cdglabs/ohm/compare/master...children-as-sequence#diff-92ae35c843ff7941815cd7b0aafac28cL61
- https://github.com/cdglabs/ohm/compare/master...children-as-sequence#diff-349b41d5ca30c7132d8b3bae3702bd29L51

Thoughts on whether or not this is a good idea? If I did this, I'd also want to implement an `asArray` method as an escape hatch.
